### PR TITLE
Task#28 - Feat: Handling quantity adjustment logic

### LINKED ISSUE #100: User Story#3: Adjust Quantity and Display Total Cost
> As a user, I want to adjust the number of items in my cart and see the total cost before proceeding to payment.

- [x] #101

- [x] When adding products already in the box, the quality of the products will also be increased.
- [x] Within the cart summary sidebar, provide controls to increase or decrease the quantity of each product.
- [x] Ensure that adjusting quantities also update the corresponding total cost for each product.
- [x] When the quantity of the product's quantity reaches 0, the product will be removed from the cart

- [x] #104

- [x] Display the calculated total cost of all products in the cart at the bottom of the sidebar.
- [x] Update the total cost dynamically as users adjust quantities.

Task#3: Test Quantity Adjustment and Total Cost Calculation

- [x] Test the quantity adjustment controls for accuracy and responsiveness.
- [x] Verify that the total cost accurately reflects the adjusted quantities.

### LINKED ISSUE #101: Task#28 - Dev Task: Implement Quantity Adjustment
- #89 
  - #100 
## OVERVIEW

- When adding a product already in the cart, increase the quantity of that product
- When the user clicks the button - it will reduce the quantity of the product
- When the user clicks the + button, the quantity of the product will be increased
- When the quantity of that product reaches 0, the product will be removed from the cart

## DEMO

https://github.com/Agility-Internship/Duong-The-Hien-ReactJS/assets/113186169/f710db71-9c13-46a5-9cb6-85a79a8b4afb

### PR DESCRIPTION
**Pull request by TheHienDuong**
- Closes #101 

## OVERVIEW
- Update addToCart, removeFromCart function
- Create new updateQuantity function
  - Increase or decrease the quantity of the product
  - Remove products when the quantity reaches 0
- Update comments and PropTypes

## DEMO

https://github.com/Agility-Internship/Duong-The-Hien-ReactJS/assets/113186169/12c6812a-dbf6-4bba-9438-bd5d0642d100

### CODE REVIEW

File: practice-two/src/App.jsx [hunk 1]
```diff
@@ -104,14 +104,21 @@ const App = () => {
...
-    setCartProductIDs((prevCartIDs) => {
-      // If the product is already in the cart, don't add it again
-      if (prevCartIDs.includes(productID)) {
-        return prevCartIDs;
+    setCartProductIDs((prevCartProducts) => {
+      const existingProduct = prevCartProducts.find((item) => item.id === productID);
+
+      if (existingProduct) {
+        setShowAddToCartModal(true);
+        // If the product already exists in the cart, update its quantity
```
Comment (phattruong-agilityio) [inline practice-two/src/App.jsx:111]: code style

File: practice-two/src/App.jsx [hunk 2]
```diff
@@ -104,14 +104,21 @@ const App = () => {
...
+        return prevCartProducts.map((item) =>
+          item.id === productID ? { ...item, quantity: item.quantity + 1 } : item,
+        );
       }
+
```
Comment (phattruong-agilityio) [inline practice-two/src/App.jsx:116]: Why `setShowAddToCartModal(true);`  twice ?

image

File: practice-two/src/App.jsx [hunk 3]
```diff
@@ -122,16 +129,30 @@ const App = () => {
...
 
+  /**
+   * Update the quantity of a product in the cart.
+   *
+   * @param {string} productID - ID of the product to update quantity for
+   * @param {number} newQuantity - New quantity value for the product
+   */
+  const updateQuantity = (productID, newQuantity) => {
+    if (newQuantity <= 0) {
+      removeFromCart(productID); // Remove the product from the cart
```
Comment (phattruong-agilityio) [inline practice-two/src/App.jsx]: code style

File: practice-two/src/components/CartItem.jsx [hunk 4]
```diff
@@ -49,24 +49,33 @@ const ProductImage = ({ src, alt }) => (
...
   const firstImage = Object.values(images)[0];
 
-  const [quantity, setQuantity] = useState(1);
+  const [quantity, setQuantity] = useState(productQuantity);
 
   const decreaseQuantity = () => {
-    if (quantity > 1) {
+    if (quantity) {
       setQuantity(quantity - 1);
+      updateQuantity(product.id, quantity - 1);
```
Comment (phattruong-agilityio) [inline practice-two/src/components/CartItem.jsx]: duplicate `quantity - 1`

File: practice-two/src/components/CartItem.jsx [hunk 5]
```diff
@@ -49,24 +49,33 @@ const ProductImage = ({ src, alt }) => (
...
+}) => {
   const { images, name, version, resolution, price } = product;
   const firstImage = Object.values(images)[0];
 
-  const [quantity, setQuantity] = useState(1);
+  const [quantity, setQuantity] = useState(productQuantity);
 
   const decreaseQuantity = () => {
-    if (quantity > 1) {
+    if (quantity) {
```
Comment (phattruong-agilityio) [inline practice-two/src/components/CartItem.jsx]: `if (quantity)` ?? Please detail the condition

File: practice-two/src/components/CartItem.jsx [hunk 6]
```diff
@@ -49,24 +49,33 @@ const ProductImage = ({ src, alt }) => (
...
       setQuantity(quantity - 1);
+      updateQuantity(product.id, quantity - 1);
     }
   };
 
   const increaseQuantity = () => {
     setQuantity(quantity + 1);
+    updateQuantity(product.id, quantity + 1);
```
Comment (phattruong-agilityio) [inline practice-two/src/components/CartItem.jsx]: duplicate `quantity + 1`

File: practice-two/src/components/ShoppingCart.jsx [hunk 7]
```diff
@@ -35,7 +41,12 @@ const ShoppingCart = ({ products = [], cartProductIDs = [], removeFromCart = ()
     document.body.classList.remove('overflow-hidden');
   };
 
-  const getProductDetails = (productId) => products.find((product) => product.id === productId);
+  const cartProducts = cartProductIDs.map((item) => ({
+    id: item.id,
+    quantity: item.quantity,
+  }));
+
```
Comment (phattruong-agilityio) [inline practice-two/src/components/ShoppingCart.jsx]: Do we need this code

```js
  const cartProducts = cartProductIDs.map((item) => ({
    id: item.id,
    quantity: item.quantity,
  }));
```

Comment (TheHienDuong) [reply practice-two/src/components/ShoppingCart.jsx]: No, they're not needed, removed

File: practice-two/src/components/CartItem.jsx (hunk 6 above)
Comment (TheHienDuong) [reply practice-two/src/components/CartItem.jsx]: Yes, updated

File: practice-two/src/components/CartItem.jsx (hunk 5 above)
Comment (TheHienDuong) [reply practice-two/src/components/CartItem.jsx]: Yes, added detail of the condition

File: practice-two/src/components/CartItem.jsx (hunk 4 above)
Comment (TheHienDuong) [reply practice-two/src/components/CartItem.jsx]: Yes,Updated

File: practice-two/src/App.jsx (hunk 3 above)
Comment (TheHienDuong) [reply practice-two/src/App.jsx]: Yes, updated

File: practice-two/src/App.jsx (hunk 2 above)
Comment (TheHienDuong) [reply practice-two/src/App.jsx:116]: setShowAddToCartModal displays a message that the user has added successfully, and in the add function there are 2 states that the product is already in the list and the product is not in the list. When successfully added, call setShowAddToCartModal to display the Notification window

**Review by phattruong-agilityio — APPROVED** [review]

**Pull request merged**